File: readAvro.scala
// /home/rstudio/spark/bin/spark-shell --repositories http://packages.confluent.io/maven/ --packages org.apache.kafka:kafka_2.11:5.4.1-ce,io.confluent:kafka-avro-serializer:5.4.1,io.confluent:kafka-schema-registry:5.4.1,org.apache.spark:spark-sql-kafka-0-10_2.11:2.4.5,org.apache.spark:spark-avro_2.11:2.4.5
// https://blog.engineering.publicissapient.fr/2017/09/27/spark-comprendre-et-corriger-lexception-task-not-serializable/package sparklyr.confluent.avro

import io.confluent.kafka.schemaregistry.client.rest.RestService
import io.confluent.kafka.serializers.KafkaAvroDeserializer
import org.apache.avro.Schema
import collection.JavaConverters._
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.Row
import io.confluent.kafka.serializers.{KafkaAvroDecoder, AbstractKafkaAvroSerDeConfig, KafkaAvroDeserializerConfig}
import java.util.Properties
import org.apache.avro.generic.GenericData

// another example
// https://github.com/jaceklaskowski/spark-structured-streaming-book/issues/1


val kafkaUrl = "broker:9092"
val schemaRegistryURL = "http://schema-registry:8081"
val topic = "parameter"
val subjectValueName = topic + "-value"

val kafkaParams = Map[String, String](
"kafka.bootstrap.servers" -> kafkaUrl,
"key.deserializer" -> "KafkaAvroDeserializer",
"value.deserializer" -> "KafkaAvroDeserializer",
"group.id" -> "structured-kafka",
"auto.offset.reset" -> "earliest",
"failOnDataLoss"-> "false",
"schema.registry.url" -> schemaRegistryURL
)

object MyDeserializerWrapper {
  val props = new Properties()
  props.put(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG, schemaRegistryURL)
  props.put(KafkaAvroDeserializerConfig.SPECIFIC_AVRO_READER_CONFIG, "true")
  val vProps = new kafka.utils.VerifiableProperties(props)
  val deser = new KafkaAvroDecoder(vProps)
  val avro_schema = new RestService(schemaRegistryURL).getLatestVersion(subjectValueName)
  val messageSchema = new Schema.Parser().parse(avro_schema.getSchema)
}

case class DeserializedFromKafkaRecord( value: String)

val spark: SparkSession = SparkSession.builder().appName("KafkaConsumerAvro").getOrCreate()

val df = spark.readStream.format("kafka").option("subscribe", topic).options(kafkaParams).load().map( x=>{
  DeserializedFromKafkaRecord(MyDeserializerWrapper.deser.fromBytes(x.getAs[Array[Byte]]("value"), MyDeserializerWrapper.messageSchema).asInstanceOf[GenericData.Record].toString)
})

df.writeStream.outputMode("append").format("console").option("truncate", false).start()

File: readAvro.R
library(sparklyr.confluent.avro)
library(sparklyr)
library(dplyr)

config <- spark_config()
config$sparklyr.shell.repositories <- "http://packages.confluent.io/maven/"
config$spark.executor.cores <- 4
kafkaUrl <- "broker:9092"
schemaRegistryUrl <- "http://schema-registry:8081"
sc <- spark_connect(master = "spark://spark-master:7077", spark_home = "spark", config=config, app_name="parameter")

stream_read_kafka_avro(sc, kafka.bootstrap.servers=kafkaUrl, schema.registry.topic="parameter", startingOffsets="latest", schema.registry.url=schemaRegistryUrl) %>%
stream_write_kafka_avro(kafka.bootstrap.servers=kafkaUrl, schema.registry.topic="output", schema.registry.url=schemaRegistryUrl, keyCols="id", mode="update") 
s <- stream_read_kafka_avro(sc, kafka.bootstrap.servers=kafkaUrl, schema.registry.topic="output", startingOffsets="latest", schema.registry.url=schemaRegistryUrl, name="output")

"select side, id from parameter" %>%
dbplyr::sql() %>%
tbl(sc, .) %>%
stream_watermark() %>%
group_by(id, time=window(timestamp, "5 minutes", "30 seconds"))%>%
summarise(n_new=n()) %>%
sdf_separate_column("time", into=c("start", "end")) %>%
select(-time) %>%
stream_write_kafka_avro(kafka.bootstrap.servers=kafkaUrl, schema.registry.topic="aggregate", schema.registry.url=schemaRegistryUrl, keyCols="id", mode="update") 

sc2 <- spark_connect(master = "spark://spark-master:7077", spark_home = "spark", config=config, app_name="aggreg")
s <- stream_read_kafka_avro(sc2, kafka.bootstrap.servers=kafkaUrl, schema.registry.topic="aggregate",
                                startingOffsets="latest", schema.registry.url=schemaRegistryUrl, name="agg") 
								
sc3 <- spark_connect(master = "local[*]", spark_home = "spark", config=config, app_name="aggreg3")
s <- stream_read_kafka_avro(sc3, kafka.bootstrap.servers=kafkaUrl, schema.registry.topic="aggregate",
                                startingOffsets="latest", schema.registry.url=schemaRegistryUrl, name="agg2") 
							








